### PR TITLE
Bids 2369/total withdrawals stats

### PR DESCRIPTION
**Pull request by LuccaBitfly**
### <samp>🤖 Generated by Copilot at 2711c31</samp>

This pull request adds support for tracking and exporting the cumulative totals of withdrawals for validators in the `validator_stats` table. It introduces a new `export-stats-totals` command to the `misc` tool, which populates the new columns `withdrawals_total` and `withdrawals_amount_total` in the table. It also improves the performance and accuracy of the `apiValidator` function and the `db` package functions that query the withdrawals totals. It includes a new SQL migration file, `temp_add_total_withdrawals_stats.sql`, to add and remove the new columns. It updates the `ValidatorStatsTableDbRow` type in the `types` package to match the new table schema.

### CODE REVIEW

File: cmd/misc/main.go [hunk 1]
```diff
@@ -783,3 +787,117 @@ func exportHistoricPrices(dayStart uint64, dayEnd uint64) {
 
 	logrus.Info("historic price update run completed")
 }
+
+func exportStatsTotals(columns string, dayStart, dayEnd, concurrency uint64) {
```
Comment (LuccaBitfly) [inline cmd/misc/main.go:791]: Example Usage:
```
go run cmd/misc/main.go --config __gitignore/config.yml --command export-stats-totals --columns withdrawals_total,withdrawals_amount_total --day-start 10 --day-end 12
```

File: db/db.go [hunk 2]
```diff
@@ -2379,12 +2379,16 @@ func GetTotalAmountWithdrawn() (sum uint64, count uint64, err error) {
...
 	err = ReaderDb.Get(&res, `
 	SELECT 
-		COALESCE(sum(w.amount), 0) as sum,
-		COALESCE(count(*), 0) as count
-	FROM blocks_withdrawals w
-	INNER JOIN blocks b ON b.blockroot = w.block_root AND b.status = '1'`)
+		COALESCE(SUM(withdrawals_amount_total), 0) as sum,
+		COALESCE(SUM(withdrawals_total), 0) as count
+	FROM validator_stats
+	WHERE day = $1`, lastExportedDay)
```
Comment (Eisei24) [inline db/db.go:2391]: **question:** What about the withdrawals of today?
Was there some agreement somewhere that they should be ignored?

File: cmd/misc/main.go [hunk 3]
```diff
@@ -783,3 +787,117 @@ func exportHistoricPrices(dayStart uint64, dayEnd uint64) {
...
+	insertQuery := fmt.Sprintf(`
+		INSERT INTO validator_stats (%s)
+		SELECT  
+			vs1.validatorindex, 
+			vs1.day,
+			%s
+		FROM validator_stats vs1 
+		LEFT JOIN validator_stats vs2 
+		ON vs2.day = vs1.day - 1 AND vs2.validatorindex = vs1.validatorindex 
+		WHERE vs1.day = $1 AND vs1.validatorindex >= $2 AND vs1.validatorindex < $3
```
[the remark this answers is not in the document]
Comment (LuccaBitfly) [reply cmd/misc/main.go:845]: Yes, if you look here
https://github.com/gobitfly/eth2-beaconchain-explorer/blob/c7f0ee883d402e7be7a39d45a064a8089eb3f265/cmd/misc/main.go#L878-L880
the parameter will never be greater than `maxValidatorIndex`

File: db/migrations/temp_add_total_withdrawals_stats.sql [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+SELECT 'up SQL query - add total withdrawal count and amount columns to stats';
+ALTER TABLE validator_stats ADD COLUMN IF NOT EXISTS withdrawals_total INT;
+ALTER TABLE validator_stats ADD COLUMN IF NOT EXISTS withdrawals_amount_total BIGINT;
+-- +goose StatementEnd
+
+-- +goose Down
+-- +goose StatementBegin
+SELECT 'down SQL query - remove total withdrawal count and amount columns from stats';
+ALTER TABLE validator_stats DROP COLUMN IF EXISTS withdrawals_total;
+ALTER TABLE validator_stats DROP COLUMN IF EXISTS withdrawals_amount;
```
[the remark this answers is not in the document]
Comment (LuccaBitfly) [reply db/migrations/temp_add_total_withdrawals_stats.sql]: fixed

File: cmd/misc/main.go [hunk 5]
```diff
@@ -783,3 +787,117 @@ func exportHistoricPrices(dayStart uint64, dayEnd uint64) {
...
+	for _, col := range columnsSlice {
+		totalClause := fmt.Sprintf("COALESCE(vs1.%s, 0) + COALESCE(vs2.%s, 0)", strings.TrimSuffix(col, "_total"), col)
+		totalClauses = append(totalClauses, totalClause)
+
+		conflictClause := fmt.Sprintf("%s = excluded.%s", col, col)
+		conflictClauses = append(conflictClauses, conflictClause)
+	}
+
+	insertQuery := fmt.Sprintf(`
+		INSERT INTO validator_stats (%s)
```
[the remark this answers is not in the document]
Comment (LuccaBitfly) [reply cmd/misc/main.go]: fixed

File: cmd/misc/main.go [hunk 6]
```diff
@@ -783,3 +787,117 @@ func exportHistoricPrices(dayStart uint64, dayEnd uint64) {
...
+
+	for day := dayStart; day <= dayEnd; day++ {
+		timeDay := time.Now()
+		logrus.Infof("exporting total sync and for columns %v for day %v", columns, day)
+
+		// get max validator index for day
+		firstEpoch, _ := utils.GetFirstAndLastEpochForDay(day + 1)
+		maxValidatorIndex, err := db.BigtableClient.GetMaxValidatorindexForEpoch(firstEpoch)
+		if err != nil {
+			utils.LogFatal(err, "error in GetAggregatedValidatorIncomeDetailsHistory: could not get max validator index from validator income history", 0, map[string]interface{}{
```
[the remark this answers is not in the document]
Comment (LuccaBitfly) [reply cmd/misc/main.go]: fixed

File: db/db.go (hunk 2 above)
Comment (LuccaBitfly) [reply db/db.go]: there was no agreement, I forgor 💀

fixed

File: cmd/misc/main.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (Eisei24) [reply cmd/misc/main.go]: The rest of the first error message then also needs to be updated.

**Review by Eisei24 — COMMENTED** [review]
Looks good, please just update the error message.

**Review by Eisei24 — APPROVED** [review]
utACK

**Review by peterbitfly — APPROVED** [review]

**Comment by LuccaBitfly** [comment]
Splitting this PR into two parts, see https://github.com/gobitfly/eth2-beaconchain-explorer/pull/2653

**Comment by LuccaBitfly** [comment]
See https://github.com/gobitfly/eth2-beaconchain-explorer/pull/2668

**Pull request closed without merging**